crates/router/src/core/recon.rs

```diff
@@ -0,0 +1,223 @@
+        merchant_id: merchant_id.clone(),
+        user_name: UserName::new(user_from_db.name)
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Failed to form username")?,
+        recipient_email: UserEmail::new(Secret::new("biz@hyperswitch.io".to_string()))
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Failed to convert recipient's email to UserEmail")?,
+        settings: state.conf.clone(),
+        subject: format!(
+            "Dashboard Pro Feature Request by {}",
+            user_from_db.email.expose().peek()
+        ),
+    };
+
+    let is_email_sent = state
+        .email_client
+        .compose_and_send_email(
+            Box::new(email_contents),
+            state.conf.proxy.https_url.as_ref(),
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Failed to compose and send email for ProFeatureRequest")
+        .is_ok();
+
+    if is_email_sent {
+        let updated_merchant_account = storage::MerchantAccountUpdate::ReconUpdate {
+            recon_status: enums::ReconStatus::Requested,
+        };
+
+        let response = db
+            .update_merchant(
+                key_manager_state,
+                merchant_account,
+                updated_merchant_account,
+                &key_store,
+            )
+            .await
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable_lazy(|| {
+                format!("Failed while updating merchant's recon status: {merchant_id:?}")
+            })?;
+
+        Ok(service_api::ApplicationResponse::Json(
+            recon_api::ReconStatusResponse {
+                recon_status: response.recon_status,
+            },
+        ))
+    } else {
+        Ok(service_api::ApplicationResponse::Json(
+            recon_api::ReconStatusResponse {
+                recon_status: enums::ReconStatus::NotRequested,
+            },
+        ))
+    }
+}
+
+pub async fn generate_recon_token(
+    state: SessionState,
+    req: ReconUser,
+) -> RouterResponse<recon_api::ReconTokenResponse> {
+    let global_db = &*state.global_store;
+    let db = &*state.store;
+    let user: UserFromStorage = global_db
+        .find_user_by_id(&req.user_id)
+        .await
+        .map_err(|e| {
+            if e.current_context().is_db_not_found() {
+                e.change_context(errors::ApiErrorResponse::InvalidJwtToken)
+            } else {
+                e.change_context(errors::ApiErrorResponse::InternalServerError)
+            }
+        })?
+        .into();
+
+    let user_role = db
+        .find_user_role_by_user_id(&req.user_id, UserRoleVersion::V1)
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::MerchantAccountNotFound)?;
+
+    core_utils::validate_profile_id_from_auth_layer(req.profile_id.clone(), &user_role)?;
```

Comment:
This will fail when user is a part of multiple profiles, and we are getting the user_role that we find first from the above code.

So, the first user role we find in the DB might not be the same profile as that of token.

Reply:
Instead of using `find_user_role_by_user_id`, use the new db functions that are added for user_role.